nifi-nar-bundles/nifi-cassandra-bundle/nifi-cassandra-distributedmapcache-service/src/main/java/org/apache/nifi/controller/cassandra/CassandraDistributedMapCache.java

```diff
@@ -0,0 +1,241 @@
+ */
+
+package org.apache.nifi.controller.cassandra;
+
+import com.datastax.driver.core.BoundStatement;
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.ResultSet;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+import org.apache.nifi.annotation.documentation.CapabilityDescription;
+import org.apache.nifi.annotation.documentation.Tags;
+import org.apache.nifi.annotation.lifecycle.OnDisabled;
+import org.apache.nifi.annotation.lifecycle.OnEnabled;
+import org.apache.nifi.cassandra.CassandraSessionProviderService;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.controller.AbstractControllerService;
+import org.apache.nifi.controller.ConfigurationContext;
+import org.apache.nifi.distributed.cache.client.Deserializer;
+import org.apache.nifi.distributed.cache.client.DistributedMapCacheClient;
+import org.apache.nifi.distributed.cache.client.Serializer;
+import org.apache.nifi.processor.util.StandardValidators;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import static org.apache.nifi.controller.cassandra.QueryUtils.createDeleteStatement;
+import static org.apache.nifi.controller.cassandra.QueryUtils.createExistsQuery;
+import static org.apache.nifi.controller.cassandra.QueryUtils.createFetchQuery;
+import static org.apache.nifi.controller.cassandra.QueryUtils.createInsertStatement;
+
+@Tags({"map", "cache", "distributed", "cassandra"})
+@CapabilityDescription("Provides a DistributedMapCache client that is based on Apache Cassandra.")
+public class CassandraDistributedMapCache extends AbstractControllerService implements DistributedMapCacheClient {
+    public static final PropertyDescriptor SESSION_PROVIDER = new PropertyDescriptor.Builder()
+            .name("cassandra-dmc-session-provider")
+            .displayName("Session Provider")
+            .description("The client service that will configure the cassandra client connection.")
+            .required(true)
+            .identifiesControllerService(CassandraSessionProviderService.class)
+            .build();
+
+    public static final PropertyDescriptor TABLE_NAME = new PropertyDescriptor.Builder()
+            .name("cassandra-dmc-table-name")
+            .displayName("Table Name")
+            .description("The name of the table where the cache will be stored.")
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_EL_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor KEY_FIELD_NAME = new PropertyDescriptor.Builder()
+            .name("cassandra-dmc-key-field-name")
+            .displayName("Key Field Name")
+            .description("The name of the field that acts as the unique key. (The CQL type should be \"blob\")")
```

Comment:
Is there a necessary restriction on the CQL type to blob? I would think TEXT, VARCHAR, or other string-based fields would be acceptable? Or is that so we can use a single serializer/deserializer?

Reply:
Yeah. TBH, we're learning Cassandra, and I didn't want to introduce a bunch of options that I couldn't immediately reason about. I figured that having a simple set of field requirements would make it easier to configure and to write a query generator that just works.

Reply:
I think it would be helpful to offer a small set of options that correspond to commonly used key/value types, such as string/text/varchar, int, blob, etc. I'm fine with having blob be the only one for now, maybe I can work on a follow-up Jira to add Key Type and Value Type dropdowns and update the doc to relax that constraint

Reply:
I'd rather do it now because my team is planning to actually use this before 1.13.0 is released, so I'd like to make sure we just git right. What types besides text and varchar should realistically be supported?

Reply:
i would think int should be supported at least for a key type